File: java-algorithm/src/main/java/com/leetcode/Problem1449.java
package com.leetcode;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * 1449. Form Largest Integer With Digits That Add up to Target
 * https://leetcode.com/problems/form-largest-integer-with-digits-that-add-up-to-target/
 *
 * cost.length == 9
 * 1 <= cost[i] <= 5000
 * 1 <= target <= 5000
 */
public class Problem1449 {
    public String largestNumber(int[] cost, int target) {
        String[] memo = new String[target + 1];
        Arrays.fill(memo, "");
        return dfs(cost, target, memo);
    }

    private String dfs(int[] cost, int target, String[] memo) {
        if (!memo[target].equals("")) return memo[target];
        if (target == 0) return "";
        memo[target] = "0";
        for (int i = 0; i < cost.length; i++) {
            if (cost[i] <= target) {
                String ret = dfs(cost, target - cost[i], memo);
                if (!ret.equals("0")) {
                    ret = (i + 1) + ret;
                    if (compare(ret, memo[target]) > 0) {
                        memo[target] = ret;
                    }
                }
            }
        }
        return memo[target];
    }

    private int compare(String s1, String s2) {
        return s1.length() == s2.length() ? s1.compareTo(s2) : s1.length() - s2.length();
    }

    private int getSize(int[] cost, int target) {
        // dp[i][j]: cost 前 i 个数，组成 j 的最大元素个数
        int[][] dp = new int[cost.length + 1][target + 1];
        for (int i = 1; i <= cost.length; i++) {
            for (int j = 1; j <= target; j++) {
                dp[i][j] = dp[i - 1][j];
                if (j == cost[i - 1])
                    dp[i][j] = Math.max(dp[i][j], 1);
                if (j > cost[i - 1] && dp[i][j - cost[i - 1]] != 0)
                    dp[i][j] = Math.max(dp[i][j], dp[i][j - cost[i - 1]] + 1);
            }
        }
        return dp[cost.length][target];
    }

    public String largestNumber2(int[] cost, int target) {
        String[] dp = new String[target + 1];
        Arrays.fill(dp, "");
        for (int i = 1; i <= target; i++) {
            dp[i] = "0";
            for (int j = 0; j < cost.length; j++) {
                if (cost[j] <= i) {
                    String ret = dp[i - cost[j]];
                    if (!ret.equals("0")) {
                        ret = (j + 1) + ret;
                        if (compare(ret, dp[i]) > 0) {
                            dp[i] = ret;
                        }
                    }
                }
            }
        }
        return dp[target];
    }

    public static void main(String[] args) {
        Problem1449 sol = new Problem1449();
        System.out.println("" + sol.largestNumber(new int[]{5,6,7,3,4,6,7,4,8}, 29));
        System.out.println("" + sol.largestNumber(new int[]{5, 4, 4, 5, 5, 5, 5, 5, 5}, 19));
        System.out.println("" + sol.largestNumber(new int[]{6,10,15,40,40,40,40,40,40}, 47));
        System.out.println("" + sol.largestNumber2(new int[]{5,6,7,3,4,6,7,4,8}, 29));
        System.out.println("" + sol.largestNumber2(new int[]{5, 4, 4, 5, 5, 5, 5, 5, 5}, 19));
        System.out.println("" + sol.largestNumber2(new int[]{6,10,15,40,40,40,40,40,40}, 47));
    }
}

File: python-algo/leetcode/problem_1449.py
"""1449. Form Largest Integer With Digits That Add up to Target
https://leetcode.com/problems/form-largest-integer-with-digits-that-add-up-to-target/
"""
from functools import lru_cache
from typing import List


class Solution:
    def largest_number(self, cost: List[int], target: int) -> str:
        def compare(a: str, b: str) -> bool:
            return a > b if len(a) == len(b) else len(a) > len(b)

        @lru_cache(None)
        def dfs(x: int) -> str:
            if x == 0:
                return ''
            res = '0'
            for i in range(len(cost)):
                if cost[i] <= x:
                    ret = dfs(x - cost[i])
                    if ret != '0':
                        ret = str(i + 1) + ret
                        if compare(ret, res):
                            res = ret
            return res

        return dfs(target)

    def largest_number3(self, cost: List[int], target: int) -> str:
        def gt(a: str, b: str) -> bool:
            return a > b if len(a) == len(b) else len(a) > len(b)

        dp = [''] * (target + 1)
        for i in range(1, target + 1):
            dp[i] = '0'
            for j in range(9):
                if cost[j] <= i:
                    ret = dp[i - cost[j]]
                    if ret != '0':
                        ret = str(j + 1) + ret
                        if gt(ret, dp[i]):
                            dp[i] = ret
        return dp[target]

    def largest_number2(self, cost: List[int], target: int) -> str:
        def get_digits() -> int:
            for i in range(1, len(cost) + 1):
                for j in range(1, target + 1):
                    dp[i][j] = dp[i - 1][j]
                    if cost[i - 1] == j:
                        dp[i][j] = max(dp[i][j], 1)
                    elif cost[i - 1] < j and dp[i][j - cost[i - 1]] != 0:
                        dp[i][j] = max(dp[i][j], 1 + dp[i][j - cost[i - 1]])
            return dp[len(cost)][target]

        dp = [[0] * (target + 1) for _ in range(len(cost) + 1)]
        digits = get_digits()
        if digits <= 0:
            return '0'
        ans = ''
        for num in range(len(cost), 0, -1):
            c = cost[num - 1]
            while target >= c and dp[-1][target] == 1 + dp[-1][target - c]:
                if target == c:
                    return ans + str(num)
                elif dp[-1][target - c] != 0:
                    ans += str(num)
                    target -= c
                else:
                    break
        return ans


if __name__ == '__main__':
    sol = Solution()
    print(sol.largest_number2([4, 3, 2, 5, 6, 7, 2, 5, 5], 9))
    print(sol.largest_number3([4, 3, 2, 5, 6, 7, 2, 5, 5], 9))
